sentry_streams/sentry_streams/pipeline/message.py

```diff
@@ -1,27 +1,140 @@
 from __future__ import annotations
 
-from dataclasses import dataclass
-from typing import (
-    Any,
-    Generic,
-    MutableSequence,
-    Optional,
-    Tuple,
-    TypeVar,
-)
-
-from sentry_kafka_schemas.codecs import Codec
-
-TIn = TypeVar("TIn")  # TODO: Consider naming this TPayload
-
-
-# A message with a generic payload
-@dataclass(frozen=True)
-class Message(Generic[TIn]):
-    payload: TIn
-    headers: MutableSequence[Tuple[str, bytes]]
-    timestamp: float
-    schema: Optional[
-        Codec[Any]
-    ]  # The schema of incoming messages. This is optional so Messages can be flexibly initialized in any part of the pipeline. We may want to change this down the road.
-    # TODO: Add support for an event timestamp
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Generic, Optional, Sequence, Tuple, TypeVar, cast
+
+from sentry_streams.rust_streams import PyAnyMessage, RawMessage
+
+TPayload = TypeVar("TPayload")
+
+
+class Message(ABC, Generic[TPayload]):
+    """
+    A generic class to represent multiple types of messages in the pipeline.
+    Streaming Steps should access the message via this class.
+
+    The actual Message classes are defined in Rust and are exported to Python
+    via pyo3. This class and its subclasses wrap the Rust classes so that we
+    can make the payload time Generic. It is not possible to create, via pyo3,
+    a class that is generic to Python nor using Rust generics.
+
+    The other reason this class exists is to have a superclass for all message
+    types in Python. In rust we have a rich enum.
+
+    TODO: Find a way to avoid redeclaring all the message types in Python.
+          we should be able to only redeclare the messages where we want to
+          make payload a Python Generic.
+    """
+
+    @property
+    @abstractmethod
+    def payload(self) -> TPayload:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def headers(self) -> Sequence[Tuple[str, bytes]]:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def timestamp(self) -> float:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def schema(self) -> str | None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def deepcopy(self) -> Message[TPayload]:
+        raise NotImplementedError
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Message):
+            return False
+        return (
+            self.payload == other.payload
+            and self.headers == other.headers
+            and self.timestamp == other.timestamp
+            and self.schema == other.schema
+        )
+
+
+class PyMessage(Generic[TPayload], Message[TPayload]):
+    """
+    A wrapper for the Rust PyAnyMessage to make the payload generic.
```

Comment:
Does that mean we'll have types like `PyMessage[IngestMetric]`? I guess I still don't fully understand how adding this wrapper class helps us achieve the generic